### PR TITLE
PKG: Add package lnn

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
towards https://github.com/lcompilers/lpython/issues/992

This PR adds a basic neural network package consisting of a single perceptron node. It can do binary classification.

**Example 1: Nand Classification, 4 data points**
```bash
weights = [0.00000000000000000e+00, 0.00000000000000000e+00, 0.00000000000000000e+00(bias)]
learn_rate = 5.00000000000000028e-02
accuracy = 0.00000000000000000e+00
epochs_cnt = 0
=================================
weights = [-1.00000000000000006e-01, -1.00000000000000006e-01, 1.00000000000000006e-01(bias)]
learn_rate = 5.00000000000000028e-02
accuracy = 1.00000000000000000e+02
epochs_cnt = 2
```
lpy_tmp

**Example 2: Randomly constructed binary classifiable example, 5 data points**
```bash
weights = [0.00000000000000000e+00, 0.00000000000000000e+00, 0.00000000000000000e+00(bias)]
learn_rate = 5.00000000000000028e-02
accuracy = 0.00000000000000000e+00
epochs_cnt = 0
=================================
weights = [-2.20000000000000029e-01, 1.00000000000000006e-01, 1.00000000000000006e-01(bias)]
learn_rate = 5.00000000000000028e-02
accuracy = 1.00000000000000000e+02
epochs_cnt = 4
```
lpy_tmp 2

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
ASR Verification fails currently.

```bash
(lp) lpython$ cat integration_tests/aaaa1.py 
from tmp_pkg import A, hi

def main():
    hi()
    a: A

main()
(lp) lpython$ lpython integration_tests/aaaa1.py --show-asr --indent
(TranslationUnit
    (SymbolTable
        1
        {
            A:
                (ExternalSymbol
                    1
                    A
                    5 A
                    tmp
                    []
                    A
                    Public
                ),
            hi:
                (ExternalSymbol
                    1
                    hi
                    5 hi
                    tmp
                    []
                    hi
                    Public
                ),
            main:
                (Function
                    (SymbolTable
                        8
                        {
                            a:
                                (Variable
                                    8
                                    a
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Struct
                                        5 A
                                        []
                                    )
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    main
                    (FunctionType
                        []
                        ()
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        []
                        .false.
                    )
                    [hi]
                    []
                    [(SubroutineCall
                        1 hi
                        ()
                        []
                        ()
                    )]
                    ()
                    Public
                    .false.
                    .false.
                ),
            tmp:
                (Module
                    (SymbolTable
                        5
                        {
                            A:
                                (StructType
                                    (SymbolTable
                                        6
                                        {
                                            x:
                                                (Variable
                                                    6
                                                    x
                                                    []
                                                    Local
                                                    ()
                                                    ()
                                                    Default
                                                    (Integer 4 [])
                                                    Source
                                                    Public
                                                    Required
                                                    .false.
                                                )
                                        })
                                    A
                                    []
                                    [x]
                                    Source
                                    Public
                                    .false.
                                    .false.
                                    ()
                                    ()
                                ),
                            hi:
                                (Function
                                    (SymbolTable
                                        7
                                        {
                                            
                                        })
                                    hi
                                    (FunctionType
                                        []
                                        ()
                                        Source
                                        Implementation
                                        ()
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        .false.
                                        []
                                        []
                                        .false.
                                    )
                                    []
                                    []
                                    [(Print
                                        ()
                                        [(StringConstant
                                            "hi"
                                            (Character 1 2 () [])
                                        )]
                                        ()
                                        ()
                                    )]
                                    ()
                                    Public
                                    .false.
                                    .false.
                                )
                        })
                    tmp
                    []
                    .false.
                    .false.
                ),
            tmp_pkg:
                (Module
                    (SymbolTable
                        3
                        {
                            A:
                                (ExternalSymbol
                                    3
                                    A
                                    5 A
                                    tmp
                                    []
                                    A
                                    Public
                                ),
                            hi:
                                (ExternalSymbol
                                    3
                                    hi
                                    5 hi
                                    tmp
                                    []
                                    hi
                                    Public
                                )
                        })
                    tmp_pkg
                    [tmp]
                    .false.
                    .false.
                )
        })
    [(SubroutineCall
        1 main
        ()
        []
        ()
    )]
)
ASR verify pass error: ASR verify: Struct::m_derived_type cannot point outside of its symbol table
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ubaid/Desktop/OpenSource/lpython/src/bin/lpython.cpp", line 1684
    return emit_asr(arg_file, lpython_pass_manager, runtime_library_dir,
  File "/Users/ubaid/Desktop/OpenSource/lpython/src/bin/lpython.cpp", line 194
    r = LCompilers::LPython::python_ast_to_asr(al, lm, *ast, diagnostics,
  File "/Users/ubaid/Desktop/OpenSource/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 6704
    throw LCompilersException("Verify failed");
LCompilersException: Verify failed
```

**Comment by Shaikh-Ubaid** [comment]
This is currently a WIP (Work in progress), hence the CI tests might not pass.

**Comment by Shaikh-Ubaid** [comment]
I posted some issues that I came across while working on this PR - https://github.com/lcompilers/lpython/issues/1721, https://github.com/lcompilers/lpython/issues/1722, https://github.com/lcompilers/lpython/issues/1723, https://github.com/lcompilers/lpython/issues/1724, https://github.com/lcompilers/lpython/issues/1725.

There is also another issue about importing `dataclass`, which I am hoping I fixed in [Support importing StructType](https://github.com/lcompilers/lpython/pull/1719/commits/4f5d2bc7610541ce53066d85be674c625abda6b9) and 
[ASR: Fix derived_type points outside symtab](https://github.com/lcompilers/lpython/pull/1719/commits/779ff705734fcd278c03d6c1598fedb6d915e5d8).

The initial commits perform some refactoring in the `pass_array_by_data()`. I refactored it while going through the pass to catch https://github.com/lcompilers/lpython/issues/1721. The refactoring is (mostly) done to break down large code into smaller functions, hopefully improving readability/maintainability.

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please review and share feedback.

File: src/libasr/pass/pass_array_by_data.cpp [hunk 1]
```diff
@@ -359,68 +359,75 @@ class EditProcedureCallsVisitor : public ASR::ASRPassBaseWalkVisitor<EditProcedu
...
-                            if ( v.proc2newproc.find(sym) != v.proc2newproc.end() ) {
-                                ASR::symbol_t* new_var_sym = v.proc2newproc[sym].first;
-                                ASR::expr_t* new_var = ASRUtils::EXPR(ASR::make_Var_t(al, var->base.base.loc, new_var_sym));
-                                ASR::call_arg_t new_arg;
-                                new_arg.m_value = new_var;
-                                new_arg.loc = arg.loc;
-                                new_args.push_back(al, new_arg);
-                                args_updated = true;
-                                use_original_arg = false;
+        void check_and_update_args_for_pass_arr_by_data_passed_as_callback(const T& x) {
```
[the remark this answers is not in the document]
Comment (Shaikh-Ubaid) [reply src/libasr/pass/pass_array_by_data.cpp:362]: It is a refactoring of code into a function. I named the function as `check_and_update_args_for_pass_arr_by_data_passed_as_callback` based on the operation that I thought the code was performing. I think it improves the code readibility. I think it is not towards any specific issue.

[the remark this answers is not in the document]
Comment (Shaikh-Ubaid) [reply src/libasr/pass/pass_array_by_data.cpp]: I just renamed it to `update_args_for_pass_arr_by_data_funcs_passed_as_callback()`.

**Review by certik — APPROVED** [review]
I think that this is fine. @czgdp1807 do you see any issues with the code changes?

**Comment by certik** [comment]
I am going to merge it, as it fixes an important bug (importing struct). Thank you @Shaikh-Ubaid, great work!

**Comment by Shaikh-Ubaid** [comment]
Thank you for the approval and merging!

**Pull request merged**